packages/react-dom/src/test-utils/ReactTestUtils.js

```diff
@@ -390,58 +438,120 @@ const ReactTestUtils = {
 
   Simulate: null,
   SimulateNative: {},
-
-  act(callback: () => void): Thenable {
+  act(callback: () => void | Thenable) {
     if (actContainerElement === null) {
-      // warn if we can't actually create the stub element
       if (__DEV__) {
-        warningWithoutStack(
-          typeof document !== 'undefined' &&
-            document !== null &&
-            typeof document.createElement === 'function',
-          'It looks like you called TestUtils.act(...) in a non-browser environment. ' +
-            "If you're using TestRenderer for your tests, you should call " +
-            'TestRenderer.act(...) instead of TestUtils.act(...).',
-        );
+        // warn if we're trying to use this in something like node (without jsdom)
+        if (didWarnAboutActInNodejs === false) {
+          didWarnAboutActInNodejs = true;
+          warningWithoutStack(
+            typeof document !== 'undefined' && document !== null,
+            'It looks like you called ReactTestUtils.act(...) in a non-browser environment. ' +
+              "If you're using TestRenderer for your tests, you should call " +
+              'ReactTestRenderer.act(...) instead of ReactTestUtils.act(...).',
+          );
+        }
       }
-      // then make it
+      // now make the stub element
       actContainerElement = document.createElement('div');
     }
-
-    const result = ReactDOM.unstable_batchedUpdates(callback);
     // note: keep these warning messages in sync with
     // createReactNoop.js and ReactTestRenderer.js
-    if (__DEV__) {
-      if (result !== undefined) {
-        let addendum;
-        if (result !== null && typeof result.then === 'function') {
-          addendum =
-            '\n\nIt looks like you wrote ReactTestUtils.act(async () => ...), ' +
-            'or returned a Promise from the callback passed to it. ' +
-            'Putting asynchronous logic inside ReactTestUtils.act(...) is not supported.\n';
-        } else {
-          addendum = ' You returned: ' + result;
+    let thenable;
+    actedUpdates((doesHavePassiveEffects, onDone) => {
+      const result = ReactDOM.unstable_batchedUpdates(callback);
```

Comment:
If it'a already wrapped in batched updates then it shouldn't need to be wrapped in batchedUpdates, too

Reply:
Hmm? I moved the batchedUpdates call from FiberScheduler to here

Reply:
If it's already wrapped in `actedUpdates`, I meant.

> I moved the batchedUpdates call from FiberScheduler to here

Why though? Remember that eventually the entirety of `act` is going to live in ReactFiberScheduler, behind a feature flag. That's probably what I would have done in this PR, in fact: Implement `act` behind a feature flag, and turn that on for test and noop renderers. For DOM you can use a (temporary) forked implementation. Then in a subsequent PR you can create a separate build of React DOM that turns on the feature flag.

Reply:
Only for bytes, since dan and seb had mentioned they’d like to ship as little as possible in Reactdom. That’s why I’m using the hack to flush effects too; in a precious commit, most of this was in the FiberScheduler. I could move that back in and put it behind a feature flag ofc.

Reply:
unsure how to proceed here. I could either move stuff into FiberScheduler, and add some weight to ReactDOM, or keep it outside, and copy paste across the 3 targets as I do now. and maybe I should put the feature flag in a follow up PR? I can bring this up on monday in chat.